### PR TITLE
Feature/25 basic bounding volumes

### PR DESCRIPTION
**Pull request by ilia-glushchenko**

### CODE REVIEW

File: Pegasus/include/boundingvolumes.hpp [hunk 1]
```diff
@@ -162,3 +164,48 @@ class BoundingSphere
 } // namespace geometry
 } // namespace pegasus
 #endif // PEGASUS_OBB_HPP
+
```
Comment (Warezovvv) [inline Pegasus/include/boundingvolumes.hpp]: ? :D

File: Pegasus/sources/boundingvolumes.cpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+using namespace volumes;
+
+volumes::Shape::Shape(volumes::Vertices const & vertices, volumes::Faces const & indices)
+        : vertices(vertices), indices(indices)
+{
+}
+
+Eigen::Vector3f volumes::calculateMeanVertex(volumes::Shape const & shape, volumes::Indices const & indices) {
+    Eigen::Vector3f sum{ 0, 0, 0 };
+    for (auto index : indices)
```
Comment (Warezovvv) [inline Pegasus/sources/boundingvolumes.cpp]: const & ?

File: Pegasus/sources/boundingvolumes.cpp [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    uint32_t max_eigen_value_index = (eigen_values[0] < eigen_values[1] ? 1 : 0);
+    if (eigen_values[max_eigen_value_index] < eigen_values[2]) {
+        max_eigen_value_index = 2;
+    }
+    Eigen::Vector3f max_dispersion_axis = eigen_vectors.col(max_eigen_value_index).normalized();
+
+    //Find extremal points on it
+    size_t min_vertex_index = 0;
+    size_t max_vertex_index = 0;
+    for (auto face_index : indices) {
```
Comment (Warezovvv) [inline Pegasus/sources/boundingvolumes.cpp]: const &?

File: Pegasus/sources/boundingvolumes.cpp [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        for (auto vertex_index : shape.indices[face_index]) {
```
Comment (Warezovvv) [inline Pegasus/sources/boundingvolumes.cpp]: const &

File: Pegasus/sources/boundingvolumes.cpp [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+)
+{
+    Vector3 const sphereMassCenter = sphere.getCenterOfMass();
+    float sphereRadius = static_cast<float>(sphere.getRadius());
+    Eigen::Vector3f sphereCenter(static_cast<float>(sphereMassCenter.x), 
+                                 static_cast<float>(sphereMassCenter.y), 
+                                 static_cast<float>(sphereMassCenter.z));
+
+    //Find point outside of the sphere and resize sphere
+    for (auto face_index : indices) {
```
Comment (Warezovvv) [inline Pegasus/sources/boundingvolumes.cpp]: const &

File: Pegasus/sources/boundingvolumes.cpp [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        for (auto vertex_index : shape.indices[face_index]) {
```
Comment (Warezovvv) [inline Pegasus/sources/boundingvolumes.cpp]: const &

File: Pegasus/sources/boundingvolumes.cpp (hunk 2 above)
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Taking std::size_t by reference will result in lower performance than by value.

File: Pegasus/sources/boundingvolumes.cpp [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    Eigen::Vector3f const mean = 1.0f / (3.0f * indices.size()) * sum;
+    return mean;
+}
+
+Eigen::Matrix3f volumes::calculateCovarianceMatrix(
+        volumes::Shape const & shape, volumes::Indices const & indices, Eigen::Vector3f const & mean)
+{
+    Eigen::Matrix3f C = Eigen::Matrix3f::Zero();
+
+    for (auto index : indices)
```
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Taking std::size_t by reference will result in lower performance than by value.

File: Pegasus/sources/boundingvolumes.cpp [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+{
+    Eigen::EigenSolver<Eigen::Matrix3f> es(covariance);
+    return es.eigenvectors().real();;
+}
+
+Eigen::Matrix3f volumes::calculateExtremalVertices(
+        Eigen::Matrix3f const & eigen_vectors, volumes::Shape const & shape, volumes::Indices const & indices)
+{
+    std::vector<Eigen::Vector3f const *, Eigen::aligned_allocator<Eigen::Vector3f const*>> vertices;
+    for (auto index : indices)
```
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Taking std::size_t by reference will result in lower performance than by value.

File: Pegasus/sources/boundingvolumes.cpp [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+            for (size_t k = 0; k < 3; ++k)
+            {
+                float const v = p[j] * p[k] + q[j] * q[k] + r[j] * r[k];
+                C(j, k) += v;
+            }
+        }
+    }
+
+    float const a = 1.0f / (3.0f * indices.size());
+    for (size_t j = 0; j < 3; ++j)
```
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Ok.

File: Pegasus/sources/boundingvolumes.cpp [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+    {
+        for (size_t k = 0; k < 3; ++k)
```
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Ok.

File: Pegasus/sources/boundingvolumes.cpp [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+{
+    std::vector<Eigen::Vector3f const *, Eigen::aligned_allocator<Eigen::Vector3f const*>> vertices;
+    for (auto index : indices)
+    {
+        auto const & face = shape.indices[index];
+        vertices.insert(vertices.end(), { &shape.vertices[face[0]], &shape.vertices[face[1]], &shape.vertices[face[2]] });
+    }
+
+    Eigen::Matrix3f extremal = Eigen::Matrix3f::Zero();
+    for (size_t i = 0; i < 3; ++i)
```
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Ok.

File: Pegasus/sources/boundingvolumes.cpp [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+}
+
+volumes::Vertices
+obb::OrientedBoundingBox::calculateBoxVertices(
+    Eigen::Matrix3f const & extremal_points, Eigen::Matrix3f const & eigen_vectors) const
+{
+    Vertices box(8, Eigen::Vector3f{ 0, 0, 0 });
+
+    std::array<Plane, 6> planes;
+    for (size_t i = 0; i < 6; ++i)
```
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Ok.

File: Pegasus/sources/boundingvolumes.cpp [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+
+void aabb::AxisAlignedBoundingBox::calculateExtremalVetices(
+        volumes::Shape const & shape, volumes::Indices const & indices, aabb::AxisAlignedBoundingBox::Box & box)
+{
+    using namespace std::placeholders;
+
+    //ToDo: optimize it, currently O(n), should be possible to do O(log(n))
+    std::vector<Eigen::Vector3f const *> valid_vertices;
+    valid_vertices.reserve(indices.size() * 3);
+    for(auto face_index : indices)
```
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Taking std::size_t by reference will result in lower performance than by value.

File: Pegasus/sources/boundingvolumes.cpp [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+    {
+        for (auto vertex_index : shape.indices[face_index])
```
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Taking std::size_t by reference will result in lower performance than by value.

File: Pegasus/sources/boundingvolumes.cpp (hunk 3 above)
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Taking std::size_t by reference will result in lower performance than by value.

File: Pegasus/sources/boundingvolumes.cpp (hunk 4 above)
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Taking std::size_t by reference will result in lower performance than by value.

File: Pegasus/sources/boundingvolumes.cpp (hunk 5 above)
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Taking std::size_t by reference will result in lower performance than by value.

File: Pegasus/sources/boundingvolumes.cpp (hunk 6 above)
Comment (ilia-glushchenko) [inline Pegasus/sources/boundingvolumes.cpp]: Taking std::size_t by reference will result in lower performance than by value.

**Pull request merged**